### DISCUSSION TITLE
nanochat working on a rtx3060 !

### DISCUSSION THREAD

**Discussion by FiveTechSoft** [karpathy/nanochat#176] [Show and tell]
(.venv) anto@Anto-iMac:~/nanochat$ PYTHONPATH="$(pwd):$PYTHONPATH" uv run torchrun --standalone --nproc_per_node=1 scripts/base_train.py \
  --depth=10 --device_batch_size=4 --max_seq_len=1024 \
  --num_iterations=800

                                                   █████                 █████
                                                  ░░███                 ░░███
 ████████    ██████   ████████    ██████   ██████  ░███████    ██████   ███████
░░███░░███  ░░░░░███ ░░███░░███  ███░░███ ███░░███ ░███░░███  ░░░░░███ ░░░███░
 ░███ ░███   ███████  ░███ ░███ ░███ ░███░███ ░░░  ░███ ░███   ███████   ░███
 ░███ ░███  ███░░███  ░███ ░███ ░███ ░███░███  ███ ░███ ░███  ███░░███   ░███ ███
 ████ █████░░████████ ████ █████░░██████ ░░██████  ████ █████░░████████  ░░█████
░░░░ ░░░░░  ░░░░░░░░ ░░░░ ░░░░░  ░░░░░░   ░░░░░░  ░░░░ ░░░░░  ░░░░░░░░    ░░░░░

Overriding: depth = 10
Overriding: device_batch_size = 4
Overriding: max_seq_len = 1024
Overriding: num_iterations = 800
Autodetected device type: cuda
/home/anto/nanochat/.venv/lib/python3.10/site-packages/torch/__init__.py:1617: UserWarning: Please use the new API settings to control TF32 behavior, such as torch.backends.cudnn.conv.fp32_precision = 'tf32' or torch.backends.cuda.matmul.fp32_precision = 'ieee'. Old settings, e.g, torch.backends.cuda.matmul.allow_tf32 = True, torch.backends.cudnn.allow_tf32 = True, allowTF32CuDNN() and allowTF32CuBLAS() will be deprecated after Pytorch 2.9. Please see https://pytorch.org/docs/main/notes/cuda.html#tensorfloat-32-tf32-on-ampere-and-later-devices (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:80.)
  _C._set_float32_matmul_precision(precision)
2025-10-26 10:01:07,805 - nanochat.common - INFO - Distributed world size: 1
Vocab size: 65,536
num_layers: 10
model_dim: 640
num_heads: 5
num_kv_heads: 5
Tokens / micro-batch / rank: 4 x 1024 = 4,096
Tokens / micro-batch: 4,096
Total batch size 524,288 => gradient accumulation steps: 128
Number of parameters: 133,038,080
Estimated FLOPs per token: 6.252134e+08
Using user-provided number of iterations: 800
Total number of training tokens: 419,430,400
Tokens : Params ratio: 3.15
Total training FLOPs estimate: 2.622335e+17
Scaling the LR for the AdamW parameters ∝1/√(640/768) = 1.095445
Muon: Grouping 40 params of shape torch.Size([640, 640]), device cuda:0, dtype torch.float32
Muon: Grouping 10 params of shape torch.Size([640, 2560]), device cuda:0, dtype torch.float32
Muon: Grouping 10 params of shape torch.Size([2560, 640]), device cuda:0, dtype torch.float32
[rank0]:W1026 10:01:14.754000 9459 .venv/lib/python3.10/site-packages/torch/_inductor/utils.py:1558] [0/0] Not enough SMs to use max_autotune_gemm mode
Step 00000 | Validation bpb: 3.3123
step 00000/00800 (0.00%) | loss: 11.090355 | lrm: 1.00 | dt: 55259.76ms | tok/sec: 74 | mfu: 0.60 | total time: 0.00m
step 00001/00800 (0.12%) | loss: 10.874537 | lrm: 1.00 | dt: 16429.54ms | tok/sec: 249 | mfu: 2.02 | total time: 0.00m
step 00002/00800 (0.25%) | loss: 10.477944 | lrm: 1.00 | dt: 16500.82ms | tok/sec: 248 | mfu: 2.01 | total time: 0.00m
step 00003/00800 (0.38%) | loss: 9.860694 | lrm: 1.00 | dt: 16547.52ms | tok/sec: 247 | mfu: 2.00 | total time: 0.00m
step 00004/00800 (0.50%) | loss: 9.276738 | lrm: 1.00 | dt: 16549.24ms | tok/sec: 247 | mfu: 2.00 | total time: 0.00m
step 00005/00800 (0.62%) | loss: 8.834044 | lrm: 1.00 | dt: 16550.21ms | tok/sec: 247 | mfu: 2.00 | total time: 0.00m
step 00006/00800 (0.75%) | loss: 8.524522 | lrm: 1.00 | dt: 16891.83ms | tok/sec: 242 | mfu: 1.96 | total time: 0.00m
step 00007/00800 (0.88%) | loss: 8.195137 | lrm: 1.00 | dt: 16583.34ms | tok/sec: 246 | mfu: 2.00 | total time: 0.00m
step 00008/00800 (1.00%) | loss: 7.973590 | lrm: 1.00 | dt: 16618.45ms | tok/sec: 246 | mfu: 1.99 | total time: 0.00m
step 00009/00800 (1.12%) | loss: 7.747945 | lrm: 1.00 | dt: 16617.96ms | tok/sec: 246 | mfu: 1.99 | total time: 0.00m
step 00010/00800 (1.25%) | loss: 7.536078 | lrm: 1.00 | dt: 16622.64ms | tok/sec: 246 | mfu: 1.99 | total time: 0.00m
step 00011/00800 (1.38%) | loss: 7.394103 | lrm: 1.00 | dt: 16617.29ms | tok/sec: 246 | mfu: 1.99 | total time: 0.28m
step 00012/00800 (1.50%) | loss: 7.320959 | lrm: 1.00 | dt: 16622.08ms | tok/sec: 246 | mfu: 1.99 | total time: 0.55m
step 00013/00800 (1.62%) | loss: 7.215820 | lrm: 1.00 | dt: 16622.72ms | tok/sec: 246 | mfu: 1.99 | total time: 0.83m
step 00014/00800 (1.75%) | loss: 7.143379 | lrm: 1.00 | dt: 16621.11ms | tok/sec: 246 | mfu: 1.99 | total time: 1.11m
step 00015/00800 (1.88%) | loss: 7.056562 | lrm: 1.00 | dt: 16619.07ms | tok/sec: 246 | mfu: 1.99 | total time: 1.39m
step 00016/00800 (2.00%) | loss: 6.953691 | lrm: 1.00 | dt: 16625.13ms | tok/sec: 246 | mfu: 1.99 | total time: 1.66m
step 00017/00800 (2.12%) | loss: 6.872063 | lrm: 1.00 | dt: 16619.11ms | tok/sec: 246 | mfu: 1.99 | total time: 1.94m
step 00018/00800 (2.25%) | loss: 6.823595 | lrm: 1.00 | dt: 16622.33ms | tok/sec: 246 | mfu: 1.99 | total time: 2.22m
step 00019/00800 (2.38%) | loss: 6.775950 | lrm: 1.00 | dt: 16621.29ms | tok/sec: 246 | mfu: 1.99 | total time: 2.49m
step 00020/00800 (2.50%) | loss: 6.755485 | lrm: 1.00 | dt: 16623.41ms | tok/sec: 246 | mfu: 1.99 | total time: 2.77m
step 00021/00800 (2.62%) | loss: 6.705476 | lrm: 1.00 | dt: 16622.92ms | tok/sec: 246 | mfu: 1.99 | total time: 3.05m
step 00022/00800 (2.75%) | loss: 6.617852 | lrm: 1.00 | dt: 16622.97ms | tok/sec: 246 | mfu: 1.99 | total time: 3.32m
step 00023/00800 (2.88%) | loss: 6.566742 | lrm: 1.00 | dt: 16618.54ms | tok/sec: 246 | mfu: 1.99 | total time: 3.60m
step 00024/00800 (3.00%) | loss: 6.539116 | lrm: 1.00 | dt: 16619.41ms | tok/sec: 246 | mfu: 1.99 | total time: 3.88m
step 00025/00800 (3.12%) | loss: 6.480117 | lrm: 1.00 | dt: 16622.90ms | tok/sec: 246 | mfu: 1.99 | total time: 4.16m
step 00026/00800 (3.25%) | loss: 6.450654 | lrm: 1.00 | dt: 16626.18ms | tok/sec: 246 | mfu: 1.99 | total time: 4.43m
step 00027/00800 (3.38%) | loss: 6.398592 | lrm: 1.00 | dt: 16628.42ms | tok/sec: 246 | mfu: 1.99 | total time: 4.71m
step 00028/00800 (3.50%) | loss: 6.376236 | lrm: 1.00 | dt: 16618.19ms | tok/sec: 246 | mfu: 1.99 | total time: 4.99m
step 00029/00800 (3.62%) | loss: 6.325626 | lrm: 1.00 | dt: 16627.56ms | tok/sec: 246 | mfu: 1.99 | total time: 5.26m
step 00030/00800 (3.75%) | loss: 6.337911 | lrm: 1.00 | dt: 16620.53ms | tok/sec: 246 | mfu: 1.99 | total time: 5.54m
step 00031/00800 (3.88%) | loss: 6.304930 | lrm: 1.00 | dt: 16632.10ms | tok/sec: 246 | mfu: 1.99 | total time: 5.82m
step 00032/00800 (4.00%) | loss: 6.279291 | lrm: 1.00 | dt: 16618.73ms | tok/sec: 246 | mfu: 1.99 | total time: 6.09m
step 00033/00800 (4.12%) | loss: 6.254782 | lrm: 1.00 | dt: 16623.47ms | tok/sec: 246 | mfu: 1.99 | total time: 6.37m
step 00034/00800 (4.25%) | loss: 6.228437 | lrm: 1.00 | dt: 16622.49ms | tok/sec: 246 | mfu: 1.99 | total time: 6.65m
step 00035/00800 (4.38%) | loss: 6.200771 | lrm: 1.00 | dt: 16628.86ms | tok/sec: 246 | mfu: 1.99 | total time: 6.93m
step 00036/00800 (4.50%) | loss: 6.173006 | lrm: 1.00 | dt: 16623.94ms | tok/sec: 246 | mfu: 1.99 | total time: 7.20m
step 00037/00800 (4.62%) | loss: 6.123704 | lrm: 1.00 | dt: 16629.04ms | tok/sec: 246 | mfu: 1.99 | total time: 7.48m
step 00038/00800 (4.75%) | loss: 6.105469 | lrm: 1.00 | dt: 16619.63ms | tok/sec: 246 | mfu: 1.99 | total time: 7.76m
step 00039/00800 (4.88%) | loss: 6.030816 | lrm: 1.00 | dt: 16627.83ms | tok/sec: 246 | mfu: 1.99 | total time: 8.03m
step 00040/00800 (5.00%) | loss: 6.015225 | lrm: 1.00 | dt: 16623.02ms | tok/sec: 246 | mfu: 1.99 | total time: 8.31m
step 00041/00800 (5.12%) | loss: 6.008800 | lrm: 1.00 | dt: 16633.89ms | tok/sec: 246 | mfu: 1.99 | total time: 8.59m
step 00042/00800 (5.25%) | loss: 5.972865 | lrm: 1.00 | dt: 16627.86ms | tok/sec: 246 | mfu: 1.99 | total time: 8.87m
step 00043/00800 (5.38%) | loss: 5.977058 | lrm: 1.00 | dt: 16625.61ms | tok/sec: 246 | mfu: 1.99 | total time: 9.14m
step 00044/00800 (5.50%) | loss: 5.953300 | lrm: 1.00 | dt: 16626.51ms | tok/sec: 246 | mfu: 1.99 | total time: 9.42m
step 00045/00800 (5.62%) | loss: 5.927799 | lrm: 1.00 | dt: 16632.79ms | tok/sec: 246 | mfu: 1.99 | total time: 9.70m
step 00046/00800 (5.75%) | loss: 5.900624 | lrm: 1.00 | dt: 16626.10ms | tok/sec: 246 | mfu: 1.99 | total time: 9.97m
step 00047/00800 (5.88%) | loss: 5.879824 | lrm: 1.00 | dt: 16635.18ms | tok/sec: 246 | mfu: 1.99 | total time: 10.25m
step 00048/00800 (6.00%) | loss: 5.853616 | lrm: 1.00 | dt: 16626.25ms | tok/sec: 246 | mfu: 1.99 | total time: 10.53m
step 00049/00800 (6.12%) | loss: 5.849559 | lrm: 1.00 | dt: 16633.60ms | tok/sec: 246 | mfu: 1.99 | total time: 10.81m
step 00050/00800 (6.25%) | loss: 5.806351 | lrm: 1.00 | dt: 16625.06ms | tok/sec: 246 | mfu: 1.99 | total time: 11.08m
keeps working...

**Comment by FiveTechSoft** [comment]
step 00080/00800 (10.00%) | loss: 5.321157 | lrm: 1.00 | dt: 16635.11ms | tok/sec: 246 | mfu: 1.99 | total time: 19.40m
step 00081/00800 (10.12%) | loss: 5.314687 | lrm: 1.00 | dt: 16631.84ms | tok/sec: 246 | mfu: 1.99 | total time: 19.68m
step 00082/00800 (10.25%) | loss: 5.285933 | lrm: 1.00 | dt: 16637.18ms | tok/sec: 246 | mfu: 1.99 | total time: 19.95m
step 00083/00800 (10.38%) | loss: 5.272572 | lrm: 1.00 | dt: 16627.03ms | tok/sec: 246 | mfu: 1.99 | total time: 20.23m
step 00084/00800 (10.50%) | loss: 5.258427 | lrm: 1.00 | dt: 16633.48ms | tok/sec: 246 | mfu: 1.99 | total time: 20.51m
step 00085/00800 (10.62%) | loss: 5.261037 | lrm: 1.00 | dt: 16628.49ms | tok/sec: 246 | mfu: 1.99 | total time: 20.78m
step 00086/00800 (10.75%) | loss: 5.259584 | lrm: 1.00 | dt: 16635.10ms | tok/sec: 246 | mfu: 1.99 | total time: 21.06m
step 00087/00800 (10.88%) | loss: 5.224455 | lrm: 1.00 | dt: 16642.50ms | tok/sec: 246 | mfu: 1.99 | total time: 21.34m

**Reply by aifartist** [reply]
@FiveTechSoft   You managed to train the 100M from scratch in 1-2 hours on a 3060!?!?!?

I was just happy to get Karpathy's full training example in the README to train on my dual 5090 setup.  I didn't bother to let it complete given my estimate was something like 60(?) hours.  Currently I'm working on getting full torch.compile max-autotune working with CUDAGraphs when micro-batch grad accumulations are being done.   I believe I can get a large performance boost from this.

**Reply by aifartist** [reply]
@FiveTechSoft   You managed to train the 100M from scratch in 1-2 hours on a 3060!?!?!?

I was just happy to get Karpathy's full training example in the README to train on my dual 5090 setup.  I didn't bother to let it complete given my estimate was something like 60(?) hours.  Currently I'm working on getting full torch.compile max-autotune working with CUDAGraphs when micro-batch grad accumulations are being done.   I believe I can get a large performance boost from this.

**Comment by FiveTechSoft** [comment]
https://forums.fivetechsupport.com/viewtopic.php?p=282103#p282103
